src/resolution/resolver.ts

```diff
@@ -51,103 +51,146 @@ const _resolveRequest = <T>(requestScope: interfaces.RequestScope) =>
-            result = binding.implementationType as unknown as T;
-        } else if (binding.type === BindingTypeEnum.DynamicValue && binding.dynamicValue !== null) {
-            result = invokeFactory(
-                "toDynamicValue",
-                binding.serviceIdentifier,
-                () => (binding.dynamicValue as (context: interfaces.Context) => any)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Factory && binding.factory !== null) {
-            result = invokeFactory(
-                "toFactory",
-                binding.serviceIdentifier,
-                () => (binding.factory as interfaces.FactoryCreator<any>)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Provider && binding.provider !== null) {
-            result = invokeFactory(
-                "toProvider",
-                binding.serviceIdentifier,
-                () => (binding.provider as interfaces.Provider<any>)(request.parentContext)
-            );
-        } else if (binding.type === BindingTypeEnum.Instance && binding.implementationType !== null) {
-            result = resolveInstance(
-                binding,
-                binding.implementationType,
-                childRequests,
-                _resolveRequest(requestScope)
-            );
-        } else {
-            // The user probably created a binding but didn't finish it
-            // e.g. container.bind<T>("Something"); missing BindingToSyntax
-            const serviceIdentifier = getServiceIdentifierAsString(request.serviceIdentifier);
-            throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifier}`);
-        }
+const _tryGetFromScope = <T>(
+    requestScope: interfaces.RequestScope,
+    binding:interfaces.Binding<T>): T | Promise<T> | null => {
 
-        if (isPromise(result)) {
-            result = result.then((resolved) => _onActivation(request, binding, resolved));
-        } else {
-            result = _onActivation(request, binding, result);
-        }
+    if (_isSingletonScope(binding) && binding.activated) {
+        return binding.cache!;
+    }
 
-        // store in cache if scope is singleton
-        if (isSingleton) {
-            binding.cache = result;
-            binding.activated = true;
+    if (
+        _isRequestScope(binding) &&
+        requestScope !== null &&
+        requestScope.has(binding.id)
+    ) {
+        return requestScope.get(binding.id);
+    }
+    return null;
+}
+
+const _isSingletonScope = (binding:interfaces.Binding<any>): boolean => {
+    return binding.scope === BindingScopeEnum.Singleton;
+}
 
-            if (isPromise(result)) {
-                result = result.catch((ex) => {
-                    // allow binding to retry in future
-                    binding.cache = null;
-                    binding.activated = false;
+const _isRequestScope = (binding:interfaces.Binding<any>): boolean => {
```

Comment:
We prefer `unknown` over `any`
```suggestion
const _isRequestScope = (binding:interfaces.Binding<unknown>): boolean => {
```

Reply:
Done